recalbox/recalbox-configgen
# Runtest

Hi guys,
I was so bored with joysticks tests, I decided to improve a bit the test framework.

To run it from the project's root, type :
`python -m runtest discover -v`
To get help : (https://docs.python.org/2/library/unittest.html) ;-) 
`python -m runtest --help`

You can also run it from any sub-directory (except /tests) to reduce the scope as soon as the root directory is in the PYTHONPATH.  

Inside :
- configgen/runtest : The runtest module with inittest override.
- configgen/tests/generators/linapple : linapple-pie tests and samples to move or develop your own tests using the new framework.
- configgen/tests/resources/ : Last version of es_inputs.cfg with 34 controllers.
- configgen/tests/resources/linapple/ : linapple-pie resources needed to run the tests.

What's new :
- Run on Ubuntu OOB !
- Test name must start with test_*
- Tests outputs are now grouped in the tmp directory of the project root.
- Each module can have its own sub-directory inside.
- Global path variables are overridden to be able to load configuration files from the tests/resources directory.
- Main function can directly be called by the test.
- The last but not the least  ... test fixtures are automatically generated. For instance with 34 Controllers, 1 to 5 players, runtest generates 160 tests.

How to use :
- Create the result method of your FixtureJoystick subclass, to register expected results. (see configgen/tests/generators/linapple/linapple_fixture.py)
- Write the test method decorated with @runtest.fixture_joystick(...) to test that expected results matched with real results. (see configgen/tests/generators/linapple/test_linappleConfig.py)
- Just 2 Methods for 160 tests ...  
- for test_main (see configgen/tests/generators/linapple/test_linappleGenerator.py)

Enjoy ;-)

PS : It's a first draft, needed to be fully documented, comments are welcome.
